orchagent/debugcounterorch.h

```diff
@@ -0,0 +1,96 @@
+    task_process_status uninstallDebugCounter(const std::string& counter_name);
+    task_process_status addDropReason(const std::string& counter_name, const std::string& drop_reason);
+    task_process_status removeDropReason(const std::string& counter_name, const std::string& drop_reason);
+
+    // Free Table Management Functions
+    void addFreeCounter(const std::string& counter_name, const std::string& counter_type);
+    void deleteFreeCounter(const std::string& counter_name);
+    void addFreeDropReason(const std::string& counter_name, const std::string& drop_reason);
+    void deleteFreeDropReason(const std::string& counter_name, const std::string& drop_reason);
+    void reconcileFreeDropCounters(const std::string& counter_name);
+
+    // Flex Counter Management Functions
+    CounterType getFlexCounterType(const std::string& counter_type);
+    void installDebugFlexCounters(
+            const std::string& counter_type,
+            const std::string& counter_stat);
+    void uninstallDebugFlexCounters(
+            const std::string& counter_type,
+            const std::string& counter_stat);
+
+    // Debug Counter Initialization Helper Functions
+    std::string getDebugCounterType(const std::vector<swss::FieldValueTuple>& values) const;
+    void createDropCounter(
+            const std::string& counter_name,
+            const std::string& counter_type,
+            const std::unordered_set<std::string>& drop_reasons);
+
+    // Debug Counter Configuration Helper Functions
+    void parseDropReasonUpdate(
+            const std::string& key,
+            const char delimeter,
+            std::string *counter_name,
+            std::string *drop_reason) const;
+    bool isDropReasonValid(const std::string& drop_reason) const;
+
+    // Data Members
+    std::shared_ptr<swss::DBConnector> m_stateDb = nullptr;
+    std::shared_ptr<swss::Table> m_debugCapabilitiesTable = nullptr;
+
+    std::shared_ptr<swss::DBConnector> m_countersDb = nullptr;
+    std::shared_ptr<swss::Table> m_counterNameToPortStatMap = nullptr;
+    std::shared_ptr<swss::Table> m_counterNameToSwitchStatMap = nullptr;
+
+    FlexCounterStatManager flex_counter_manager;
+
+    std::unordered_map<std::string, std::unique_ptr<DebugCounter>> debug_counters;
+
+    // free_drop_counters are drop counters that have been created by a user
+    // that do not have any drop reasons associated with them yet. Because
+    // we cannot create a drop counter without any drop reasons, we keep track
+    // of these counters in this table.
+    std::unordered_map<std::string, std::string> free_drop_counters;
+
+    // free_drop_reasons are drop reasons that have been added by a user
+    // that do not have a counter associated with them yet. Because we
+    // cannot add drop reasons to a counter that doesn't exist yet,
+    // we keep track of the reasons in this table.
+    std::unordered_map<std::string, std::unordered_set<std::string>> free_drop_reasons;
+};
```

Comment:
Question, looks to me that free_drop_counters and free_drop_reasons are counter and reason waiting to be paired with each other. Is this designed so that user can create them in any order?

What if user add, say 2 drop reasons for a same counter,  and the 2 drop reasons were both added after counter, so the operation would like:

add_counter(c);
add_counter(c):
add_reason(r1);
add_reason(r2);

Can you add 2 identical counters to the free_counters? If not, then would r2 be left never be paired?

Reply:
Once r1 is added then it will be reconciled against c and a drop counter will be created. When we add r2 we will first check and see that the counter exists, then the reason will be added to the counter directly.